core-bundle/src/Security/DataContainer/DataContainerSubject.php

```diff
@@ -0,0 +1,58 @@
+<?php
+
+declare(strict_types=1);
+
+/*
+ * This file is part of Contao.
+ *
+ * (c) Leo Feyer
+ *
+ * @license LGPL-3.0-or-later
+ */
+
+namespace Contao\CoreBundle\Security\DataContainer;
+
+class DataContainerSubject
+{
+    private string $table;
+    private int|string|null $id;
+    private array $attributes;
+
+    public function __construct(string $table, int|string|null $id = null, array $attributes = [])
+    {
+        $this->table = $table;
+        $this->id = $id;
+        $this->attributes = $attributes;
+    }
+
+    public function __toString(): string
+    {
+        $subject = [];
+        $subject[] = 'Table: '.$this->getTable();
+
+        if ($id = $this->getId()) {
+            $subject[] = 'ID: '.$id;
+        }
+
+        if ($attributes = $this->getAttributes()) {
+            $subject[] = 'Attributes: '.implode(', ', $attributes);
```

Comment:
this should also output the `$attributes` keys?

Reply:
What if they do not have keys? What if they are objects? Not sure if we can fix this really.

Reply:
if they are objects, we can't really implode it either?

Reply:
I know. The question is what to do.

Reply:
maybe dump JSON?

Reply:
Wouldn't work with objects either but maybe we can just ignore that. It's only used as debugging message anyway.

Reply:
I will use JSON and mark this as resolved. Otherwise I'm losing track here.